### PR TITLE
[Antithesis] Transient Rows Workflow

### PR DESCRIPTION
**Pull request by jeremyk-91**
## General
**Before this PR**:
- We don't really have a workflow that involves short lived _rows_.

**After this PR**:

==COMMIT_MSG==
Introduce a transient row workflow for our workload server. This executes transactions that write a primary table row and an index row (which uses columns to identify the rows), then deletes the immediately preceding row.
==COMMIT_MSG==

**Priority**: P2

**Concerns / possible downsides (what feedback would you like?)**:
- Don't really like that this hotspots, is that a problem for the things we want to verify?

**Is documentation needed?**: No.

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**: Yes, I intentionally broke some API but since it's workload it should be fine

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**: No

## Testing and Correctness
**What was existing testing like? What have you done to improve it?**: New code is tested

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**: No

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**: No

## Execution
Workload is not deployed in production.

## Scale
Workload is not deployed in production.

## Development Process
**Where should we start reviewing?**: TransientRowsWorkflows

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**: This enables a new workflow. Not worth investment IMO.

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                    if (!violations.isEmpty()) {
+                        log.error(
+                                "Detected rows which did not agree with state in a summary index: {}",
+                                SafeArg.of("violations", violations));
+                    }
+                })
+                .build();
+    }
+
+    private static Invariant<List<CrossCellInconsistency>> createSummaryRowInvariant(
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java:105]: Could we push this into it's own separate class?

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        return store.readWrite(txn -> {
+            String tableName = configuration.tableConfiguration().tableName();
+            txn.write(tableName, ImmutableWorkloadCell.of(SUMMARY_ROW, taskIndex), VALUE);
+            txn.write(tableName, ImmutableWorkloadCell.of(taskIndex, COLUMN), VALUE);
+            if (taskIndex > 0) {
+                int indexToRead = SECURE_RANDOM.nextInt(taskIndex);
+                txn.read(tableName, ImmutableWorkloadCell.of(SUMMARY_ROW, indexToRead));
+                txn.read(tableName, ImmutableWorkloadCell.of(indexToRead, COLUMN));
+
+                txn.delete(tableName, ImmutableWorkloadCell.of(SUMMARY_ROW, taskIndex - 1));
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java:85]: We probably should check if the value exists _before_ issuing a delete; in production, I'd expect us to almost always to only delete rows that we know exist!

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/SingleRowTwoCellsWorkflowConfiguration.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
 @JsonDeserialize(as = ImmutableSingleRowTwoCellsWorkflowConfiguration.class)
 @JsonTypeName(SingleRowTwoCellsWorkflowConfiguration.TYPE)
 public interface SingleRowTwoCellsWorkflowConfiguration extends WorkflowConfiguration {
-    String TYPE = "singleRowTwoCells";
+    String TYPE = "single-row-two-cells";
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/SingleRowTwoCellsWorkflowConfiguration.java:30]: nice

Comment (jeremyk-91) [reply atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/SingleRowTwoCellsWorkflowConfiguration.java:30]: yeah not sure how I missed it the first time around

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        return history.stream()
+                .map(WitnessedTransaction::actions)
+                .flatMap(TransientRowsWorkflows::findInconsistencyInTransactionActions)
+                .collect(Collectors.toList());
+    }
+
+    private static Stream<CrossCellInconsistency> findInconsistencyInTransactionActions(
+            List<WitnessedTransactionAction> actions) {
+        Set<WitnessedReadTransactionAction> readTransactionActions = actions.stream()
+                .filter(action -> action instanceof WitnessedReadTransactionAction)
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java:132]: `WitnessedReadTransactionAction.class::isInstance`

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java (hunk 1 above)
Comment (Sam-Kramer) [reply atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java:108]: Ok let's keep it as is

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+                                TransientRowsWorkflows.VALUE));
+    }
+
+    @Test
+    public void subsequentTransactionsWriteThenReadThenDelete() {
+        WorkflowHistory history = transientRowsWorkflow.run();
+        List<WitnessedTransaction> witnessedTransactions = history.history();
+        for (int index = 1; index < ITERATION_COUNT; index++) {
+            assertThat(witnessedTransactions.get(index).actions()).hasSize(6).satisfies(actions -> {
+                assertThat(actions.subList(0, 2)).allMatch(action -> action instanceof WitnessedWriteTransactionAction);
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java:111]: `WitnessedWriteTransactionAction.class::isInstance`

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+                                                TransientRowsWorkflows.SUMMARY_ROW, ITERATION_COUNT - 2)),
+                                Optional.empty())
+                        .build()));
+        verifyNoMoreInteractions(violationConsumer);
+    }
+
+    @SuppressWarnings("unchecked") // Mocking of a known type
+    @Test
+    public void invariantReportsViolationsFromTransactionHistory() {
+        Consumer<List<CrossCellInconsistency>> violationConsumer = mock(Consumer.class);
```
Comment (Sam-Kramer) [inline atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java:179]: Any reason to not just use an `AtomicReference::set`?

File: atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflows.java:132]: yep, nice

File: atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java (hunk 6 above)
Comment (jeremyk-91) [reply atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java:179]: I did think about that, but wanted to verify the consumer is only invoked once (and it felt like the instrumentation of managing a list of lists to check that we don't call it more than once was overkill). What do you think?

Comment (Sam-Kramer) [reply atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java:179]: maybe a half-way compromise would be to spy `AtomicReference`?

Comment (Sam-Kramer) [reply atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java:178]: I think you updated it everywhere but here :D

Comment (jeremyk-91) [reply atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/TransientRowsWorkflowsTest.java]: you're right! fixed

**Review by Sam-Kramer — APPROVED** [review]

**Pull request merged**